File: 20-validParenthesis/isValid.js
"use strict";
/* // imperative, with a stack

function isValid(s: string): boolean {
    if (s.length == 0) {
        return true;
    }
    if (s.length % 2 !== 0) {
        return false;
    }
    const stack : string[] = [];
    let ptr: number = 0;
    while (ptr < s.length) {
        const top = stack[stack.length - 1];
        if (closes(top, s[ptr])) {
            stack.pop();
        }
        else {
            stack.push(s[ptr]);
        }
        ptr++;
    }
    return stack.length === 0;
}
*/
// functional, with recursion
// This actually performs better than the imperative version
function isValid(s) {
    if (s.length == 0) {
        return true;
    }
    if (s.length % 2 !== 0) {
        return false;
    }
    return solveFrom(s, 0) == s.length;
}
function solveFrom(s, ptr) {
    if (ptr >= s.length) {
        return -1; // error.
    }
    const initial = s[ptr];
    if (initial === '(' || initial === '[' || initial === '{') {
        const final = solveFrom(s, ptr + 1);
        if (final === -1) {
            return -1;
        }
        if (closes(initial, s[final])) {
            if (final === s.length - 1) {
                return s.length;
            }
            return solveFrom(s, final + 1);
        }
    }
    return ptr;
}
function closes(a, b) {
    return (a === '(' && b === ')')
        || (a === '[' && b === ']')
        || (a === '{' && b === '}');
}
// tests
console.log(isValid('()')); // true
console.log(isValid('()[]{}')); // true
console.log(isValid('(]')); // false
console.log(isValid('([)]')); // false

File: 20-validParenthesis/isValid.hs
isValid :: String -> Bool
isValid [] = True
isValid (p : ps) = isOpen p && closes p p' && isValid ps'
  where
    (p' : ps') = getNext ps

getNext :: String -> String
getNext [] = ""
getNext (p : ps) = if isOpen p && closes p p' then getNext ps' else p : ps
  where
    (p' : ps') = getNext ps

isOpen :: Char -> Bool
isOpen '(' = True
isOpen '[' = True
isOpen '{' = True
isOpen _ = False

closes :: Char -> Char -> Bool
closes '(' ')' = True
closes '[' ']' = True
closes '{' '}' = True
closes _ _ = False

-- tests

test1 = isValid "()" == True
test2 = isValid "([])" == True
test3 = isValid "([)]" == False
test4 = isValid "([{}])" == True

-- main
main :: IO ()
main = do
  print test1
  print test2
  print test3
  print test4